src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java

```diff
@@ -579,76 +592,157 @@ public void concreteIndicesWildcardExpansion() {
         assertThat(newHashSet(md.concreteIndices(IndicesOptions.fromOptions(true, true, true, true), "testX*")), equalTo(newHashSet("testXXX", "testXXY", "testXYY")));
     }
 
+    /**
+     * test resolving _all pattern (null, empty arry or "_all") for random IndicesOptions
+     */
+    @Test
+    public void concreteIndicesAllPatternRandom() {
+        for (int i = 0; i < 10; i++) {
+            String[] allIndices = null;
+            switch (randomIntBetween(0, 2)) {
+            case 0:
+                break;
+            case 1:
+                allIndices = new String[0];
+                break;
+            case 2:
+                allIndices = new String[] { MetaData.ALL };
+                break;
+            }
+
+            IndicesOptions indicesOptions = IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(), randomBoolean());
+            MetaData metadata = MetaData.builder().build();
+
+            // with no indices, asking for all indices should return empty list or exception, depending on indices options
+            if (indicesOptions.allowNoIndices()) {
+                String[] concreteIndices = metadata.concreteIndices(indicesOptions, allIndices);
+                assertThat(concreteIndices, notNullValue());
+                assertThat(concreteIndices.length, equalTo(0));
+            } else {
+                try {
+                    metadata.concreteIndices(indicesOptions, allIndices);
+                    fail("calling concreteIndices for _all indices with allowNoIndices=false should throw exception");
+                } catch (ElasticsearchIllegalArgumentException | IndexMissingException e) {
```

Comment:
I chose to throw the ElasticsearchIllegalArgumentException when the wildcard-expansion options are all off (expandWildcardsOpen, expandWildcardsClosed), allowNoIndices is also off and the method call is made with empty or null indices-names list. The IndexMissingException is thrown when wildcard expansion happend but returned nothing (and allowNoIndices is off).

Should I change this to IndexMissingException in the first case also? Feels to me more like a mistake by the callern when he requests "_all" by way of null or empty list but doesn't turn wildcard expansion on.